pkg/app/pipedv1/plugin/kubernetes/deployment/server.go

```diff
@@ -206,28 +206,29 @@ func (a *DeploymentService) loadManifests(ctx context.Context, deploy *model.Dep
 	return manifests, nil
 }
 
-func (a *DeploymentService) ExecuteStage(ctx context.Context, request *deployment.ExecuteStageRequest) (*deployment.ExecuteStageResponse, error) {
+func (a *DeploymentService) ExecuteStage(ctx context.Context, request *deployment.ExecuteStageRequest) (response *deployment.ExecuteStageResponse, _ error) {
+	lp := a.logPersister.StageLogPersister(request.GetInput().GetDeployment().GetId(), request.GetInput().GetStage().GetId())
+	defer func() {
+		// When the piped cancelled the RPC while the stage is still runnning, we should not mark the log persister as completed.
+		// we use the context error to check if the RPC is cancelled by the piped.
+		// we don't use the response error to check context cancellation because the response error can be set without wrapping the context error.
+		if !response.GetStatus().IsCompleted() && errors.Is(ctx.Err(), context.Canceled) {
+			return
+		}
```

Comment:
I realized that we have to handle two cancellations, but we cannot determine the cause because the cancel is propagated beyond the gRPC layer.

1. cancellation occurred by graceful shutdown.
    - in this case, we should not mark the log persister as completed.
2. cancellation occurred by the user via the web UI.
    - in this case, we should return the status as canceled and mark the log persister as completed.

Reply:
I realized the plugin receives SIGTERM or SIGINT when the graceful shutdown is going, so I implemented the handling of them.
When the cancellation occurs without receiving such signals, the log persister should be completed.